content/blog/learn-pulumi-inputs-outputs/index.md

```diff
@@ -0,0 +1,160 @@
+    return { endpoint: listener.endpoint.hostname, publicIPs: publicIps  };
+};
+```
+
+As the diagram earlier depicted, another thing we can easily do with outputs is export them. When we export an output, Pulumi captures that value and makes it easily available on the stack. This includes at the command line, in the pulumi console, and even if you want one stack to depend on another stack's output. You can see it in action with the following script:
+
+```bash
+for i in {1..25}; do curl http://$(pulumi stack output endpoint); done
+```
+
+Let's look at other ways of interacting with the concrete value of an output. In this example, we want to parse out the subdomain of the load balancer's hostname. Usually, you would expect a string, and we can call functions like substring and indexOf directly, but in this case, we can't because it's an Output. This is where we can use the apply function. Pulumi invokes a callback when it knows the concrete value and apply takes the callback and returns another output that we can export.
+
+```typescript
+export const subdomain = alb.hostname.apply(hostname => hostname.substring(0, indexOf(".")));
+```
+
+You can use apply to print to the console or allocate resources inside of an apply, but it can be dangerous because the resource may not be created during a preview and apply will never execute.
+
+Those are the basics, but before wrapping up, let's look at some convenient helper methods Pulumi offers for dealing with outputs.
+
+If all you want to do is call a method on the string or number or the underlying value, Pulumi offers a convenient way of accessing the members directly, even though it's an output. For example, we're calling the function to uppercase a string.
+
+```typescript
+export const upperHostname = alb.hostname.toUpperCase();
+```
+
+In another case, we may want to use concrete values from multiple outputs. We could do this by nesting multiple applys, but Pulumi provides `all` as a convenience function that can take multiple outputs and return a single output.
+
+```typescript
+export const url = pulumi.all([alb.hostname, alb.port]).apply( ( [hostname, port]) =>  `http://${hostname}:${port}` );
+```
+
+String concatenation is so commonplace that Pulumi offers a helper function for it. The `interpolate` function allows you to reference outputs without calling apply or all on any of
+them.
+
+``` typescript
+export const url = pulumi.interpolate`http://${alb.hostname}:${alb.port}`;
+```
+
+This article is a brief introduction to Pulumi inputs and outputs. Learn more about the Pulumi [programming model]({{< relref "/docs/intro/concepts/programming-model" >}}) by visiting Pulumi Docs.
+
+You can also watch the *Pulumi Help: Resource Inputs and Outputs (Node.js* on [PulumiTV](https://www.youtube.com/c/PulumiTV/).
+
+{{< youtube "lybOxul2otM" >}}
+
+{{% notes type="info" %}}
+Learn Pulumi is an irregular series about the Pulumi programming model.
```

Comment:
I'd probably just drop this note, but if you want to keep it, I'd suggest using a word other than "irregular" -- it reads here as "weird/odd". 😄 Maybe "new" series, if that's the intent?